test/integration/cni/pod_traffic_across_az_test.go

```diff
@@ -0,0 +1,304 @@
+
+		nodes, err := f.K8sResourceManagers.NodeManager().GetNodes(f.Options.NgNameLabelKey, f.Options.NgNameLabelVal)
+		Expect(err).ToNot(HaveOccurred())
+
+		azToPod = make(map[string]coreV1.Pod)
+
+		for i := range nodes.Items {
+			// node label key "topology.kubernetes.io/zone" is well known label populated by cloud controller manager
+			// guaranteed to be present and represent the AZ name
+			// Ref https://kubernetes.io/docs/reference/labels-annotations-taints/#topologykubernetesiozone
+			azName = nodes.Items[i].ObjectMeta.Labels["topology.kubernetes.io/zone"]
+			interfaceToPodList = common.GetPodsOnPrimaryAndSecondaryInterface(nodes.Items[i], "role", "az-test", f)
+			if len(interfaceToPodList.PodsOnSecondaryENI) > 0 {
+				azToPod[azName] = interfaceToPodList.PodsOnSecondaryENI[0]
+			}
+			if len(interfaceToPodList.PodsOnPrimaryENI) > 0 {
+				azToPod[azName] = interfaceToPodList.PodsOnPrimaryENI[0]
+			}
+		}
+	})
+
+	JustAfterEach(func() {
+		By("Deleting the Daemonset.")
+		err = f.K8sResourceManagers.DaemonSetManager().DeleteAndWaitTillDaemonSetDeleted(testDaemonSet, utils.DefaultDeploymentReadyTimeout)
+		Expect(err).ToNot(HaveOccurred())
+
+	})
+
+	Context("While testing API Server Connectivity", func() {
+		It("Should connect to the API Server", func() {
+			// Standard paths for SA token, CA cert and API Server URL
+			token_path := "/var/run/secrets/kubernetes.io/serviceaccount/token"
+			cacert := "/var/run/secrets/kubernetes.io/serviceaccount/ca.crt"
+			api_server_url := "https://kubernetes.default.svc/api"
+
+			for az := range azToPod {
+				fmt.Printf("Testing API Server Connectivity from AZ %s \n", az)
+				sa_token := []string{"cat", token_path}
+				token_value, _, err := RunCommandOnPod(azToPod[az], sa_token)
+				Expect(err).ToNot(HaveOccurred(), fmt.Sprintf("failed to get SA token for pod in %s", az))
+				bearer := fmt.Sprintf("Authorization: Bearer %s", token_value)
+				test_api_server_connectivity := []string{"curl", "--cacert", cacert, "--header", bearer, "-X", "GET",
+					api_server_url,
+				}
+
+				api_server_stdout, _, err := RunCommandOnPod(azToPod[az], test_api_server_connectivity)
+				// Descriptive error message on failure to connect to API Server from particular AZ.
+				Expect(err).ToNot(HaveOccurred(), fmt.Sprintf("Error while connecting to API Server from %s", az))
+				Expect(api_server_stdout).ToNot(BeEmpty())
+				Expect(api_server_stdout).To(ContainSubstring("APIVersions"))
+			}
+		})
+	})
+})
+
+func CheckConnectivityBetweenPods(azToPod map[string]coreV1.Pod, port int, testerExpectedStdOut string, testerExpectedStdErr string, getTestCommandFunc func(serverPod coreV1.Pod, port int) []string) {
+
+	By("checking connection on same node, primary to primary")
+
+	for az1 := range azToPod {
+		for az2 := range azToPod {
+			if az1 != az2 {
+				fmt.Printf("Testing Connectivity from Pod IP1 %s (%s) to Pod IP2 %s (%s) \n",
+					azToPod[az1].Status.PodIP, az1, azToPod[az2].Status.PodIP, az2)
```

Comment:
Do we capture on which AZ to AZ the communication is broken?

Reply:
Can we also monitor latency/RTT? Maybe as a first cut we can just monitor if the connectivity time between pods in a particular AZ is increasing maybe from past 3-4 runs?